src/crimson/os/seastore/seastore.cc

```diff
@@ -400,23 +400,214 @@ seastar::future<store_statfs_t> SeaStore::stat() const
+    } else {
+      temp = false;
+      ceph_assert(start >= range_start && start < range_end);
+    }
+  }
+  if (end.hobj.is_max()) {
+    if (temp) {
+      ranges.emplace_back(start, range_temp_end);
+      ranges.emplace_back(range_start, range_end);
+    } else {
+      ranges.emplace_back(start, range_end);
+    }
+  } else {
+    if (end.hobj.is_temp()) {
+      if (temp) {
+        ranges.emplace_back(start, end);
+      } else {
+        ceph_assert(0 == "impossible");
+      }
+    } else {
+      if (temp) {
+        ranges.emplace_back(start, range_temp_end);
+        ranges.emplace_back(range_start, end);
+      } else {
+        ranges.emplace_back(start, end);
+      }
+    }
+  }
+  return ranges;
+}
+
 seastar::future<std::tuple<std::vector<ghobject_t>, ghobject_t>>
 SeaStore::list_objects(CollectionRef ch,
-                        const ghobject_t& start,
-                        const ghobject_t& end,
-                        uint64_t limit) const
+                       const ghobject_t& start,
+                       const ghobject_t& end,
+                       uint64_t limit) const
 {
   using RetType = typename OnodeManager::list_onodes_bare_ret;
+  if (start.is_max() || start.hobj.is_max() || start > end) {
+    return seastar::make_ready_future<
+      std::tuple<std::vector<ghobject_t>, ghobject_t>>(
+      std::make_tuple(std::vector<ghobject_t>(), ghobject_t::get_max()));
+  }
   return seastar::do_with(
-      RetType(),
-      [this, start, end, limit] (auto& ret) {
-    return repeat_eagain([this, start, end, limit, &ret] {
+    RetType(),
+    RetType(),
+    std::list<std::pair<ghobject_t, ghobject_t>>(),
+    std::move(limit),
+    [this, ch, start, end](auto& ret, auto& pret, auto& ranges, auto& limit) {
+    return repeat_eagain([this, ch, start, end, &limit, &ret, &pret, &ranges] {
       return transaction_manager->with_transaction_intr(
         Transaction::src_t::READ,
         "list_objects",
-        [this, start, end, limit](auto &t)
+        [this, ch, start, end, &limit, &pret, &ranges](auto &t)
       {
-        return onode_manager->list_onodes(t, start, end, limit);
+        return get_coll_bits(
+          ch, t
+	  ).si_then([this, ch, &t, start, end, &limit, &pret, &ranges](auto bits) {
+          if (!bits) {
+            return OnodeManager::list_onodes_iertr::make_ready_future<
+              OnodeManager::list_onodes_bare_ret>(std::make_tuple(
+              std::vector<ghobject_t>(), ghobject_t::get_max()));
+          } else {
+            auto filter = get_objs_range(ch, *bits);
+            ranges = get_ranges(ch, start, end, filter);
+            return trans_intr::repeat([this, &t, &pret, &ranges, &limit]() mutable
+              -> OnodeManager::list_onodes_iertr::future<seastar::stop_iteration> {
+              auto ite = ranges.begin();
+              auto pstart = ite->first;
+              auto pend = ite->second;
+              return seastar::do_with(
```

Comment:
I don't think this do_with is necessary, list_onodes() appears to correctly capture its arguments as needed.